File: solutions/typescript/binary-search/1/binary-search.ts
export function find(haystack: number[], needle: number): number | never {
  return find_in_range(haystack, 0, haystack.length, needle);
}

function find_in_range(haystack: number[], begin: number, end: number, needle: number): number | never {
  if (begin == end) throw new Error("Value not in array");
  let mid = Math.trunc((begin + end) / 2);
  if (needle == haystack[mid]) 
    return mid;
  else if (needle < haystack[mid]) 
    return find_in_range(haystack, begin, mid, needle);
  else 
    return find_in_range(haystack, mid + 1, end, needle);
}

File: solutions/fsharp/binary-search/1/BinarySearch.fs
module BinarySearch

let rec find_in_range low high (input: int array) value =
    if low >= high then None
    else
        let mid = (low + high) / 2
        match input[mid] with
        | n when n = value -> Some mid
        | n when n < value -> find_in_range (mid + 1) high input value
        | _ -> find_in_range low mid input value

let find input value = 
    find_in_range 0 (Array.length input) input value
